Class based views

makkaronis4e opened the pull request:
Add class based views
Resolve #24

makkaronis4e committed:
Fix base.html

makkaronis4e committed:
Add class-based views

makkaronis4e commented:
@webknjaz done

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+
 
+class PostList(ListView):
+    """Show all your Post objects."""
```
Your? Whom is it talking about?

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+class Protected(View):
+    """Protect views that need to show only for authorised users."""
+
+    @method_decorator(login_required)
```
Did it work like this before rewrite?

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+
+
+class NewPost(FormView, Protected):
+    """Return page for adding new Post."""
```
Don't use actionable verbs in class docstring as it describes some entity, not an action.

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+
+    def post(self, request, *args, **kwargs):
+        """Get info from form and save it as 'post' object."""
         form = PostForm(request.POST)
```
You shouldn't need to initialize your own form instance. It should be pre-populated by internal CBV machinery and available as `self.form`

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+    def post(self, request, *args, **kwargs):
+        """Get info from form and save it as 'post' object."""
         form = PostForm(request.POST)
         if form.is_valid():
```
`post()` method will only be called if it's valid (based on the `form_valid()` method call earlier), so this check is also useless.

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+        form.instance.author = self.request.user
+        return super().form_valid(form)
+
+    def post(self, request, *args, **kwargs):
```
I bet `post()` method does not accept request as an argument. args/kwargs also look unrealistic

webknjaz replied:
So remove it.

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
         form = PostForm(request.POST)
         if form.is_valid():
             post = form.save(commit=False)
             post.author = request.user
```
You already do this in `form_valid()` method

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
             post.author = request.user
-            post.published_date = timezone.now()
             post.save()
             return redirect('post_detail', pk=post.pk)
```
Based on above, I'd say that it is never called. success_url is used instead.

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
             post = form.save(commit=False)
             post.author = request.user
-            post.published_date = timezone.now()
             post.save()
```
And this of course is done by CBV as well

webknjaz commented on blog/views.py:
```diff
@@ -1,53 +1,67 @@
+
+    model = Post
+    fields = ['title', 'text']
+    success_url = reverse_lazy('post_list')
```
You'd better redirect to post

makkaronis4e committed:
Rewrite view for post edit

webknjaz committed:
Fix post creation using proper CreateView

webknjaz committed:
Declare get_absolute_url on model for success URL

It makes sense to do the work in one place and not touch views
unless really needed so.

webknjaz committed:
Drop unused imports from views
